lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,370 @@
+	{
+		HornifyModule &hm = getAnalysis<HornifyModule> ();
+
+		//load the Horn clause database
+		auto &db = hm.getHornClauseDB ();
+		db.buildIndexes ();
+
+		//build the wto
+		HornClauseDBCallGraph callgraph(db);
+		HornClauseDBWto db_wto(callgraph);
+		db_wto.buildWto();
+
+		//guess candidates
+		guessCandidate(db);
+
+		//run main algorithm
+		runOnDB(db);
+
+
+		return false;
+	}
+
+	void PredicateAbstraction::getAnalysisUsage (AnalysisUsage &AU) const
+	{
+	    AU.addRequired<HornifyModule> ();
+	    AU.setPreservesAll();
+	}
+
+	void PredicateAbstraction::runOnDB(HornClauseDB &db)
+	{
+		HornClauseDB new_DB(db.getExprFactory ());
+		int index = 0;
+
+		for(Expr rel : db.getRelations())
+		{
+			predToBiNumMap.insert(std::pair<Expr, int>(rel, index));
+			index++;
+		}
+
+		for(Expr rel : db.getRelations())
+		{
+			outs() << "OLD PRED: " << *rel << "\n";
+			ExprVector new_args;
+			//Push fdecl name
+			new_args.push_back(bind::fname(rel));
+			//Push boolean types
+			for(int i=0; i < db.getRelations().size(); i++)
+			{
+				new_args.push_back(mk<BOOL_TY>(rel->efac()));
+			}
+			Expr new_rel = mknary<FDECL>(new_args);
+			outs() << "NEW PRED: " << *new_rel << "\n";
+			new_DB.registerRelation(new_rel);
+			oldToNewPredMap.insert(std::pair<Expr, Expr>(rel, new_rel));
+		}
+		outs() << "NEW DB: \n";
+		outs() << new_DB << "\n";
+
+		for(HornClauseDB::RuleVector::iterator it = db.getRules().begin(); it != db.getRules().end(); ++it)
+		{
+			HornRule r = *it;
+			outs() << "OLD RULE HEAD: " << *(r.head()) << "\n";
+			outs() << "OLD RULE BODY: " << *(r.body()) << "\n";
+
+			//Map from predicate apps to bool vars
+			std::map<Expr, Expr> predAppToBoolVarMap;
+
+			//Count occurrence time for each relation in per rule
+			std::map<Expr, int> relOccurrenceTimesMap;
+			ExprVector pred_vector;
+			get_all_pred_apps(r.body(), db, std::back_inserter(pred_vector));
+			pred_vector.push_back(r.head());
+			for(ExprVector::iterator it = pred_vector.begin(); it!= pred_vector.end(); ++it)
+			{
+				if(relOccurrenceTimesMap.find(bind::fname(*it)) == relOccurrenceTimesMap.end())
+				{
+					relOccurrenceTimesMap.insert(std::pair<Expr, int>(bind::fname(*it), 1));
+				}
+				else
+				{
+					relOccurrenceTimesMap[bind::fname(*it)] += 1;
+				}
+			}
+
+			//construct new rule head.
+			Expr rule_head = r.head();
+			Expr new_rule_head_rel = oldToNewPredMap.find(bind::fname(rule_head))->second;
+
+			ExprVector new_rule_head_args;
+			for(std::map<Expr, int>::iterator it = predToBiNumMap.begin(); it != predToBiNumMap.end(); ++it)
+			{
+				if(it->first == bind::fname(rule_head))
+				{
+					int occur_num = relOccurrenceTimesMap[bind::fname(rule_head)];
+					assert(occur_num >= 1);
+					int bool_var_index = it->second;
+					std::ostringstream oss;
+					oss << "b" << bool_var_index;
+					for(int i=0; i<occur_num - 1; i++)
+					{
+						oss << "\'";
+					}
+					Expr bi = bind::boolVar(mkTerm<std::string>(oss.str(), db.getExprFactory ()));
```

Comment:
Avoid using string names. Use structured expressions instead. In this case, you probably want to use `variant::variant` to create a variant of a boolean variable.

Name the boolean variables based on the name of the predicate. This will be easier to debug. You can number them based on their occurrence in the rule (i.e., first occurrence is 0, second occurrence is 1, etc. Independently of the name).